### PR TITLE
Feature/ami

### PR DESCRIPTION
**Pull request by miguelmalvarez**
This is a PR to add Adjusted_Mutual_Information (http://en.wikipedia.org/wiki/Adjusted_mutual_information), a quite common way to evaluate clustering, given a gold standard. Some of the initial values for testing, and the main assumptions, are based on the sklearn implementation. The PR also includes some auxiliary functions that we have refactored into the utils. 

I would love to hear your opinion if the auxiliary functions should stay in the metrics ns or not, and also if the name "metrics" is meaningful enough (again, based on sklearn) or not.

### CODE REVIEW

File: test/clj/clojurewerkz/statistiker/metrics_test.clj [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    (is (almost= (mutual_information [1 2 3] [1 2 3]) 1.0986122886681096 tol))
+    (is (= (mutual_information [1 1 1] [1 1 1]) 0.0))
+    (is (= (mutual_information [1 2 3 4] [1 1 1 1]) 0.0))
+    (is (= (mutual_information [:a :a :a :a] [1 1 1 1]) 0.0))
+    (is (almost= (mutual_information [:a 6 :d :f] [1 2 3 4]) 1.3862943611198906 tol))
+    (is (almost= (mutual_information [0 0 :c2 :c1] [0 0 87 99]) 1.0397207708399179 tol))
+    (is (almost= (mutual_information [0 0 'docid2' 'docid'] [0 99 87 0]) 0.6931471805599453 tol))
+    (is (almost= (mutual_information [0 1 2 0] [0 1 2 3]) 1.0397207708399179 tol))))
+
+(deftest adjusted_mutual_information-values
```
Comment (michaelklishin) [inline test/clj/clojurewerkz/statistiker/metrics_test.clj:37]: Please use standard Clojure function naming conventions: that is, `dash-separators`.

File: test/clj/clojurewerkz/statistiker/metrics_test.clj [hunk 2]
```diff
@@ -0,0 +1,49 @@
+(ns clojurewerkz.statistiker.metrics-test
+  (:require [clojurewerkz.statistiker.metrics :refer :all]
+            [clojurewerkz.statistiker.utils :refer [almost=]]
+            [clojure.test :refer :all]))
+; Tolerance for values
+(def tol 1e-8)
+
+(deftest wrong-input-mi
+  (testing "invalid inputs for mutual_information"
+    (is (thrown? AssertionError (mutual_information [1 2 3] [2 3])))
```
Comment (michaelklishin) [inline test/clj/clojurewerkz/statistiker/metrics_test.clj:10]: If you have a bunch of `clojure.test/is` macros with basically identical structure you may want to consider using `clojure.test/are`.

**Comment by ifesdjeen** [comment]
I'm still reviewing the change (I trust it works I just want to learn a bit more on ADI, since we didn't use it). Please give me some time (max a day or two) to understand it all better. 

On a side node I've been thinking of adding an Elbow Method and "Rule Of Thumb", do you think we could put them all into some namespace and use together?

**Comment by ifesdjeen** [comment]
Hey @miguelmalvarez @dcorney, I'm going to go ahead and merge it, and will make some stylistic changes (to make code look more like the rest of the library), if you don't mind. I'll get back to you on style asap, so that you could review it all.

Maybe let's chat some day about your current needs, since we have a couple of features in pipeline, and you seem to be quite actively using it. Maybe we could adjust our course to match your needs better?

File: test/clj/clojurewerkz/statistiker/metrics_test.clj (hunk 1 above)
Comment (miguelmalvarez) [inline test/clj/clojurewerkz/statistiker/metrics_test.clj:37]: Sorry, too many brain-switches between Python and Clojure :)

**Comment by miguelmalvarez** [comment]
@ifesdjeen, @michaelklishin Thanks for the speed and sorry about the style mistakes. I would love to have a chat and know what other features you guys have in the pipeline, please feel free to drop me an email.

The Elbow Method and the "Rule Of Thumb" can be seen as utility/tuning functions for clustering, while the AMI is an evaluation metric. I would suggest to have eval.clj and util.clj inside clustering. What do you think?

**Comment by ifesdjeen** [comment]
Sounds good. Let me finish restyling and we'll move them around.

**Comment by ifesdjeen** [comment]
@miguelmalvarez also, thanks a ton for such a neat feature, must have been a bunch of work.

Also, I've started digging into the code. There are some things I've changed (no semantical changes, everything works exactly the same though), but those are mostly performance optimizations.

**Comment by miguelmalvarez** [comment]
Fantastic, thank you guys for all the good work. I will be more careful with the style in future PRs.
Looking forward to hear about the new stuff you are working on.

**Pull request merged**